File: contest-management-java/src/main/java/repository/database/CompetitionDbRepository.java
package repository.database;

import model.Competition;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.CompetitionRepository;
import utils.JdbcUtils;

import java.sql.*;
import java.util.*;

public class CompetitionDbRepository implements CompetitionRepository {
    private JdbcUtils dbUtils;

    private static final Logger logger = LogManager.getLogger();

    public CompetitionDbRepository(Properties props) {
        logger.info("Initialising CompetitionDbRepository with properties: {}", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Iterable<Competition> findAll() {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        List<Competition> competitions = new ArrayList<>();

        try (PreparedStatement prepStmt = conn.prepareStatement("select * from Competitions")) {
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("id");
                    String competitionType = result.getString("competition_type");
                    String ageCategory = result.getString("age_category");
                    Competition comp = new Competition(competitionType, ageCategory);
                    comp.setId(id);
                    competitions.add(comp);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB" + e);
        }
        logger.traceExit();
        return competitions;
    }

    @Override
    public Competition find(Long aLong) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        Competition comp = null;
        try (PreparedStatement prepStmt = conn.prepareStatement("select * from Competitions where id = ?")) {
            prepStmt.setLong(1, aLong);
            try (ResultSet result = prepStmt.executeQuery()) {
                if (result.next()) {
                    long id = result.getLong("id");
                    String competitionType = result.getString("competition_type");
                    String ageCategory = result.getString("age_category");
                    comp = new Competition(competitionType, ageCategory);
                    comp.setId(id);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB" + e);
        }
        logger.traceExit();
        return comp;
    }

    @Override
    public Long add(Competition entity) {
        logger.traceEntry("saving task {}", entity);
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "insert into Competitions " +
                        "(competition_type, age_category) " +
                        "values (?, ?)", Statement.RETURN_GENERATED_KEYS)) {
            prepStmt.setString(1, entity.getCompetitionType());
            prepStmt.setString(2, entity.getAgeCategory());

            int result = prepStmt.executeUpdate();
            ResultSet resultSet = prepStmt.getGeneratedKeys();
            if (resultSet.next())
                return resultSet.getLong(1);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public void remove(Long aLong) {
        logger.traceEntry("removing task {}", aLong);
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn.prepareStatement("delete from Competitions where id = ?")) {
            prepStmt.setLong(1, aLong);
            int result = prepStmt.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(Competition entity) {
        logger.traceEntry("updating task {}", entity);
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn
                .prepareStatement("update Competitions set competition_type = ?, age_category = ? where id = ?")) {
            prepStmt.setString(1, entity.getCompetitionType());
            prepStmt.setString(2, entity.getAgeCategory());
            prepStmt.setLong(3, entity.getId());
            int result = prepStmt.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public int countParticipantsForCompetition(Long competitionId) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        int count = 0;
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select cnt from " +
                        "(select id, competition_type, age_category, count(competition_id) cnt " +
                        "from Competitions C " +
                        "left join Registrations R on C.id = R.competition_id " +
                        "group by competition_id, competition_type, age_category)" +
                        "where id = ?")) {
            prepStmt.setLong(1, competitionId);
            try (ResultSet result = prepStmt.executeQuery()) {
                if (result.next())
                    count = result.getInt("cnt");
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
        return count;
    }

    @Override
    public Iterable<String> findTypesByString(String competitionType) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        List<String> competitionTypes = new ArrayList<>();

        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select distinct competition_type " +
                        "from Competitions " +
                        "where competition_type like '%' || ? || '%'")) {
            prepStmt.setString(1, competitionType);
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    String compType = result.getString("competition_type");
                    competitionTypes.add(compType);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB" + e);
        }
        logger.traceExit();
        return competitionTypes;
    }

    @Override
    public Iterable<String> findAgeCategories(String competitionType) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        List<String> ageCategories = new ArrayList<>();

        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select age_category " +
                        "from Competitions " +
                        "where competition_type = ?")) {
            prepStmt.setString(1, competitionType);
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    String ageCategory = result.getString("age_category");
                    ageCategories.add(ageCategory);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB" + e);
        }
        logger.traceExit();
        return ageCategories;
    }

    @Override
    public Competition findByProps(String competitionType, String ageCategory) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        Competition c = null;
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select * from Competitions " +
                        "where competition_type = ? and age_category = ?")) {
            prepStmt.setString(1, competitionType);
            prepStmt.setString(2, ageCategory);
            try (ResultSet result = prepStmt.executeQuery()) {
                if (result.next()) {
                    Long id = result.getLong("id");
                    String compType = result.getString("competition_type");
                    String ageCat = result.getString("age_category");
                    c = new Competition(compType, ageCat);
                    c.setId(id);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB" + e);
        }
        logger.traceExit();
        return c;
    }
}

File: contest-management-cs/contest-management-gui-cs/repository/database/CompetitionDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using contest_management_gui_cs.model;
using contest_management_gui_cs.utils;
using log4net;

namespace contest_management_gui_cs.repository
{
    public class CompetitionDbRepository : ICompetitionRepository
    {
        private readonly DbUtils _dbUtils;
        
        private static readonly ILog Log = LogManager.GetLogger("mylogger");

        public CompetitionDbRepository(string connectionName)
        {
            _dbUtils = new DbUtils(connectionName);
        }
        
        public IEnumerable<Competition> FindAll()
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand("select * from Competitions", conn);
            try
            {
                using var reader = cmd.ExecuteReader();
                var competitions = new List<Competition>();
                while (reader.Read())
                {
                    competitions.Add(new Competition
                    {
                        Id = reader.GetInt64(0),
                        CompetitionType = reader.GetString(1),
                        AgeCategory = reader.GetString(2)
                    });
                }

                return competitions;
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return new List<Competition>();
        }

        public Competition Find(long id)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select * from Competitions
                 where id = @id",
                conn);
            cmd.Parameters.AddWithValue("@id", id);
            try
            {
                using var reader = cmd.ExecuteReader();
                if (reader.Read())
                    return new Competition
                    {
                        Id = reader.GetInt64(0),
                        CompetitionType = reader.GetString(1),
                        AgeCategory = reader.GetString(2),
                    };
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return default;
        }

        public void Add(Competition entity)
        {
            Log.TraceEntry(entity.ToString());

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"insert into Competitions (competition_type, age_category) VALUES
                (@competition_type, @age_category)",
                conn);
            cmd.Parameters.AddWithValue("@competition_type", entity.CompetitionType);
            cmd.Parameters.AddWithValue("@age_category", entity.AgeCategory);
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB " + e);
            }
            finally
            {
                Log.TraceExit();
            }
        }

        public void Remove(long id)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand("delete from Competitions where id = @id", conn);
            cmd.Parameters.AddWithValue("@id", id);
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB " + e);
            }
            finally
            {
                Log.TraceExit();
            }
        }

        public void Update(Competition entity)
        {
            Log.TraceEntry(entity.ToString());

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"update Competitions set
                competition_type = @competition_type,
                age_category = @age_category
                where id = @id",
                conn);
            cmd.Parameters.AddWithValue("@competition_type", entity.CompetitionType);
            cmd.Parameters.AddWithValue("@age_category", entity.AgeCategory);
            cmd.Parameters.AddWithValue("@id", entity.Id);
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB " + e);
            }
            finally
            {
                Log.TraceExit();
            }
        }

        public IDictionary<Competition, int> CountParticipantsForEachCompetition()
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select C.id, competition_type, age_category, count(competition_id) cnt from Competitions C
                left join Registrations R on C.id = R.competition_id
                group by competition_id, competition_type, age_category",
                conn);
            try
            {
                using var reader = cmd.ExecuteReader();
                var dict = new Dictionary<Competition, int>();
                while (reader.Read())
                {
                    dict[new Competition
                    {
                        Id = reader.GetInt64(0),
                        CompetitionType = reader.GetString(1),
                        AgeCategory = reader.GetString(2)
                    }] = reader.GetInt32(3);
                }

                return dict;
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB " + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return new Dictionary<Competition, int>();
        }

        public IEnumerable<string> FindTypesByString(string competitionType)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select distinct competition_type from Competitions
                where competition_type like '%' || @competition_type || '%'",
                conn);
            cmd.Parameters.AddWithValue("@competition_type", competitionType);
            try
            {
                using var reader = cmd.ExecuteReader();
                var competitionTypes = new List<string>();
                while (reader.Read())
                {
                    competitionTypes.Add(reader.GetString(0));
                }

                return competitionTypes;
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return new List<string>();
        }

        public IEnumerable<string> FindAgeCategories(string competitionType)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select age_category from Competitions
                where competition_type = @competition_type",
                conn);
            cmd.Parameters.AddWithValue("@competition_type", competitionType);
            try
            {
                using var reader = cmd.ExecuteReader();
                var ageCategories = new List<string>();
                while (reader.Read())
                {
                    ageCategories.Add(reader.GetString(0));
                }

                return ageCategories;
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return new List<string>();
        }

        public Competition FindByProps(string competitionType, string ageCategory)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select * from Competitions
                where competition_type = @competition_type and age_category = @age_category",
                conn);
            cmd.Parameters.AddWithValue("@competition_type", competitionType);
            cmd.Parameters.AddWithValue("@age_category", ageCategory);
            try
            {
                using var reader = cmd.ExecuteReader();
                var competitions = new List<Competition>();
                if (reader.Read())
                    return new Competition
                    {
                        Id = reader.GetInt64(0),
                        CompetitionType = reader.GetString(1),
                        AgeCategory = reader.GetString(2)
                    };
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return default;
        }
    }
}